### PR TITLE
New package: singularity-3.7.1

### PR DESCRIPTION
**Pull request by olafmersmann**
#### General
- [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the package proposed for inclusion on a regular basis and confirm it works for me

The tests require `sudo`, so they are disabled. To start an interactive Alpine container, run installation 
```
$ singularity pull alpine
$ singularity run alpine_latest.sif
``` 

#### Does it build and run successfully? 
- [x] I built this PR locally for my native architecture, (x86_64)

### CODE REVIEW

File: srcpkgs/singularity/template [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Template file for 'singularity'
+pkgname=singularity
+version=3.7.1
+revision=1
+wrksrc="${pkgname}"
+build_style=go
```
Comment (ericonr) [inline srcpkgs/singularity/template:6]: This looks like a datapoint for #26417 :p

File: srcpkgs/singularity/INSTALL [hunk 2]
```diff
@@ -0,0 +1,9 @@
+case "${ACTION}" in
+post)
+	if [ ! -d var/lib/singularity/mnt/session ]; then
+		mkdir -p var/lib/singularity/mnt/session
+	fi
```
Comment (ericonr) [inline srcpkgs/singularity/INSTALL:5]: Isn't `mkdir -p` enough here?

File: srcpkgs/singularity/INSTALL [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+	chown root:root var/lib/singularity/mnt/session
```
Comment (ericonr) [inline srcpkgs/singularity/INSTALL:6]: Why is this necessary? Root should be the owner by default.

File: srcpkgs/singularity/INSTALL [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+	chown root:users usr/libexec/singularity/bin/starter-suid
+	chmod 4755 usr/libexec/singularity/bin/starter-suid
```
Comment (ericonr) [inline srcpkgs/singularity/INSTALL:8]: Shouldn't this be 4750? That's what xbps-uchroot uses, at least.

File: srcpkgs/singularity/template [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+hostmakedepends="wget pkg-config go cryptsetup"
+makedepends="libressl-devel libseccomp-devel"
+depends="squashfs-tools cryptsetup"
+short_desc="HPC centric container platform"
+maintainer="Olaf Mersmann <olafm@p-value.net>"
+license="BSD-3-Clause-LBNL"
+homepage="https://sylabs.io/singularity/"
+distfiles="https://github.com/hpcng/${pkgname}/releases/download/v${version}/${pkgname}-${version}.tar.gz"
+checksum=82d2c65063560195ec34551931be3c325b95e8e2009e92755fd7daad346e083c
+nocross=yes
```
Comment (ericonr) [inline srcpkgs/singularity/template:18]: What kind of errors did you get?

File: srcpkgs/singularity/template [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	# * Don't install bash completions into /etc/bash_completion.d
+	vsed -i \
+		-e 's@^GO_MODFLAGS := .*@GO_MODFLAGS := -mod=vendor@' \
+		-e 's@INSTALLFILES += $(bash_completion_INSTALL)@@' \
+		builddir/Makefile
+}
+
+do_build() {
+	export srcdir=$PWD
+	make -C builddir old_config=
```
Comment (ericonr) [inline srcpkgs/singularity/template:42]:
```suggestion
	srcdir=$PWD make -C builddir old_config=
```

Avoids polluting the environment :)

File: srcpkgs/singularity/template [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+do_install() {
+	make DESTDIR="${DESTDIR}" -C builddir install man
+	vlicense LICENSE.md
+	vcompletion builddir/etc/bash_completion.d/singularity bash
+	rm -fr "${DESTDIR}"/usr/lib64
```
Comment (ericonr) [inline srcpkgs/singularity/template:49]: Given that `/usr/lib64` should be a symlink to `/usr/lib`, this seems very wrong... Why is it required?

Also, I prefer to use `-r` so it errors out if the file doesn't exist, which would mean the command isn't necessary :)

File: srcpkgs/singularity/template [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+do_check() {
+	# XXX: tests require sudo, so skip for now.
+	# make -C builddir unit-test
+	# make -C builddir integration-test
+	# make -C builddir e2e-test
```
Comment (ericonr) [inline srcpkgs/singularity/template:56]:
```suggestion
	# make -C builddir unit-test integration-test e2e-test
```

For brevity

File: srcpkgs/singularity/template (hunk 1 above)
Comment (olafmersmann) [reply srcpkgs/singularity/template:6]: Yes. Similar situation to `aerc`.

File: srcpkgs/singularity/INSTALL (hunk 2 above)
Comment (olafmersmann) [reply srcpkgs/singularity/INSTALL:5]: Yes, fixed.

File: srcpkgs/singularity/INSTALL (hunk 3 above)
Comment (olafmersmann) [reply srcpkgs/singularity/INSTALL:6]: Yes, wanted to be safe. Removed it.

File: srcpkgs/singularity/INSTALL (hunk 4 above)
Comment (olafmersmann) [reply srcpkgs/singularity/INSTALL:8]: 4755 is the upstream default but 4750 works fine as well.
In that case, do we also want to add a `singularity` group to further restrict access?

File: srcpkgs/singularity/template (hunk 6 above)
Comment (olafmersmann) [reply srcpkgs/singularity/template:42]: Good point, fixed.

File: srcpkgs/singularity/INSTALL (hunk 4 above)
Comment (ericonr) [reply srcpkgs/singularity/INSTALL:8]: I assume you want 4750 since you made it belong to the `users` group, otherwise chown'ing it makes no difference that I can see.

And yes, it's likely that a `_singularity` group would be best.

File: srcpkgs/singularity/template (hunk 5 above)
Comment (olafmersmann) [reply srcpkgs/singularity/template:18]: The `makeit` configure doesn't support GO crossbuilds so it chokes wenn GOARCH is set:
```
...
 checking: host C++ compiler... c++
 checking: host Go compiler (at least version 1.13)... not found!
mconfig: could not complete configuration
```

I'd really like to avoid patching crossbuild support into their Makefiles.

File: srcpkgs/singularity/template (hunk 8 above)
Comment (olafmersmann) [reply srcpkgs/singularity/template:56]: Fixed.

File: srcpkgs/singularity/template (hunk 7 above)
Comment (olafmersmann) [reply srcpkgs/singularity/template:49]: Can't remember why I added it, it's not necessary so I removed it.

File: srcpkgs/singularity/template (hunk 5 above)
Comment (ericonr) [reply srcpkgs/singularity/template:18]: Heh, np. You could consider making changes upstream, though :P

File: srcpkgs/singularity/INSTALL (hunk 4 above)
Comment (olafmersmann) [reply srcpkgs/singularity/INSTALL]: Changed.

Comment (ericonr) [reply srcpkgs/singularity/INSTALL]: Btw, `chmod`'ing to 4750 in `do_install` is also necessary, otherwise there will be a window of time where the binary is 4755 and anyone can run anything.

Comment (olafmersmann) [reply srcpkgs/singularity/INSTALL]: Fixed.

File: srcpkgs/singularity/INSTALL [hunk 9]
```diff
@@ -0,0 +1,6 @@
+case "${ACTION}" in
+post)
```
Comment (Johnnynator) [inline srcpkgs/singularity/INSTALL:2]:
```suggestion
post)
if [ "$(readlink -f . 2>/dev/null || echo .)" != "/" ]; then
	exit 1
else
	mkdir -p var/lib/singularity/mnt/session
	chown root:_singularity usr/libexec/singularity/bin/starter-suid
	chmod 4755 usr/libexec/singularity/bin/starter-suid
fi
```
Maybe do smth like this to prevent changing the binary to a potentially wrong group when executing xbps-install with `--rootidr`

**Review by Johnnynator — CHANGES_REQUESTED** [review]

File: srcpkgs/singularity/INSTALL (hunk 9 above)
Comment (ericonr) [reply srcpkgs/singularity/INSTALL:2]: I like the idea, but I'd prefer to have it as a general helper, otherwise this will have to be repeated across a few INSTALL scripts (including XBPS; it's not a huge issue there because the gid for `xbuilder` is constant, but still not great).

The helper shouldn't be added here though, so I will do it afterwards.

Comment (olafmersmann) [reply srcpkgs/singularity/INSTALL:2]: I added the check and replaced the `mkdir` with `make_dirs` in the template.

File: srcpkgs/singularity/INSTALL [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+	if [ "$(readlink -f . 2>/dev/null || echo .)" != "/" ]; then
+		exit 1
+	else
+		chown root:_singularity usr/libexec/singularity/bin/starter-suid
+		chmod 4755 usr/libexec/singularity/bin/starter-suid
```
Comment (ericonr) [inline srcpkgs/singularity/INSTALL:7]: Isn't an explicit failure here pretty bad? I think XBPS stops updates if you die with failure (instead of returning 0).

Comment (Johnnynator) [reply srcpkgs/singularity/INSTALL:7]: True that. Didn't think of that.
